New authorization system

ares opened the pull request:
will be squash, clean and rebased later when it's ready

ares committed:
New permissions model

Contains:
* Usergroup roles support
* Usergroup nesting (with cycle prevention)
* Caching of user roles and membership
* Admin flag can be assigned to usergroup
* Role filters for better granularity
* Migration of old DB to new permission model

ohadlevy commented on app/models/role.rb:
```diff
@@ -30,7 +30,10 @@ class Role < ActiveRecord::Base
-  has_many :users, :through => :user_roles
+  has_many :users, :through => :user_roles, :source => :owner, :source_type => 'User'
+  has_many :usergroups, :through => :user_roles, :source => :owner, :source_type => 'Usergroup'
+  has_many :cached_user_roles
```
:dependent  => :destroy ? on all relationships?

ares replied:
cache should be handled by UserRole which has `ruby has_many :cached_user_roles, :dependent => :destroy` and by UsergroupMember so dependent should not be needed here

on the other hand there's no reason not to specify dependent here, it shouldn't break anything and will make sure there's no cache for deleted role, so I'll add it

others (through) relations should not be deleted when you delete role I think

ohadlevy commented on app/models/cached_usergroup_member.rb:
```diff
@@ -0,0 +1,10 @@
+  belongs_to :user
+  belongs_to :usergroup
+
+  def self.build_cache!(attributes)
```
why wrap the method?

ares replied:
I thought there will be more logic, use create! directly

ohadlevy commented on app/models/user.rb:
```diff
@@ -66,6 +79,25 @@ class User < ActiveRecord::Base
+  def self.search_by_admin(key, operator, value)
+    value      = value == 'true' ? true : false
+    conditions = [self.table_name, Usergroup.table_name].map do |base|
+      "(#{base}.admin = ?" + (value ? ')' : " OR #{base}.admin IS NULL)")
```
DRY with scope? or just use scoped_search?

ares replied:
I didn't find a way to specify scope instead of hash for scoped_search's :ext_method, @abenari is that possible? also I'm not aware of a way how scope could be converted to string to reuse admin scopes at least for the resulting hash

abenari replied:
The solution looks OK to me.
Scope is not an option here because the search language enables expressions like: "admin or name=amos", scopes can only work as 'and' condition as in is_admin.name_is("amos")
scoped search 'natural' language can only define expressions like "admin or usergroup.admin"

abenari replied:
there is a small issue with the implementation, however, the operator is ignored and "admin != true" will get the wrong results.

ares replied:
good catch, thank you, I'll fix it

ares replied:
@abenari could you please check new version? are there any other operators that I should be aware of? I couldn't find too much about this in scoped_search docs

ohadlevy replied:
see http://blog.theforeman.org/2012/01/advance-search-tips-in-foreman.htmlfor
some more examples

ohadlevy commented on app/models/usergroup.rb:
```diff
@@ -1,9 +1,18 @@
   has_many :usergroup_members, :dependent => :destroy
-  has_many :users,      :through => :usergroup_members, :source => :member, :source_type => 'User'
-  has_many :usergroups, :through => :usergroup_members, :source => :member, :source_type => 'Usergroup'
+  has_many :users,      :through => :usergroup_members, :source => :member, :source_type => 'User', :dependent => :destroy
```
Am I reading this correctly? if you delete the group you remove all of its users too?

ares replied:
nope, that's common confusion of :dependent in has_many :through 
according to documentation:

If using with the :through option, the association on the join model must be a belongs_to, and the records **which get deleted are the join records**, rather than the associated records.

http://apidock.com/rails/ActiveRecord/Associations/ClassMethods/has_many

ohadlevy replied:
thanks! would it make it more readable if we move it to the join table instead?

ares replied:
well I spent some time on this and this was the only way to force rails to trigger callbacks on join model which we need, join models are removed without instantiation if you just rely on destroy of usergroup_members only

so maybe adding a comment would help? (ofc to all other has_many through with dependent key)

ohadlevy replied:
Interesting... so this would mean we have other places in the code which
are broken?

ares replied:
I don't think it's used anywhere, usually n:m relation entities are small objects without any logic, here we must execute also destroy callbacks to destroy cache which is quite extraordinary. I just went through all models having after_destroy or before_destroy callback and no (except being added) is joining entity.

ares commented on test/factories/user_related.rb:
```diff
@@ -0,0 +1,46 @@
+FactoryGirl.define do
```
before anyone comments here - this is something to discuss and refactor, however it's pretty useful during development to have factories defined together in one file

lzap commented on app/models/usergroup_member.rb:
```diff
@@ -1,5 +1,130 @@
+# Redmine - project management software
+# Copyright (C) 2006-2009  Jean-Philippe Lang
```
I'd add normal Red Hat header and keep Jean-Philippe in the copyright holders.

ares replied:
ok, since we create a new permission system maybe it's time to remove all redmine links? do we have some policy to include license in every file? if not I'd remove that, other models does not contain it either

domcleal replied:
If this new code came from Redmine, leave this header in place.  Don't remove it from existing classes unless they've been rewritten.  No need to add a Red Hat header, the licence is stated in the readme.

ares replied:
OK, this one didn't come from Redmine, it's just being built on original system taken from Redmine so I'll remove it just from this file.

lzap replied:
Yeah we do not have the license in all files (although its recommended
practice).

But if there are any bits written by JP Lang, we should state this
somewhere, maybe in the authors list.

## 

Later,

 Lukas "lzap" Zapletal
 irc: lzap #theforeman

lzap commented on db/migrate/20131114084718_extend_user_role.rb:
```diff
@@ -0,0 +1,21 @@
+    remove_index :user_roles, :owner_type
+
+    rename_column :user_roles, :owner_id, :user_id
+    remove_column :user_roles, :owner_type
```
Missing

```
remove_column :user_roles, :owner_id
```

?

ares replied:
it's done by rename_column, however add_column in method `up` line 5 should have been removed, I'll fix this

ares replied:
already fixed

dLobatog committed:
API v2 migrated to new permissions model

API v1 migration to new permissions model

Set authorizer filter for controllers

Various typo errors on API controllers corrected

Remove role validator to make migration script pass

Authorizer in application controller

@authorizer goes in application_controller and it is called from the
views, since it's a helper method. This saves us from having an
after_filter or @authorizer defined everywhere.
However I believe this might be a security risk and Authorizer should
be a singleton class.

Use @smart_proxies over @proxies for consistency

dLobatog commented:
https://github.com/ares/foreman/pull/3 some topics for discussion

dLobatog commented on lib/tasks/migrate_permissions.rake:
```diff
@@ -0,0 +1,204 @@
+        # later for hosts
+        filters[:domains]    = user.domains.uniq.map { |cr| "id = #{cr.id}" }.join(' or ')
+
+        # hostrgroups
```
typo

dLobatog commented on lib/tasks/migrate_permissions.rake:
```diff
@@ -0,0 +1,204 @@
+        filter = filters[:facts]
+        search = user.facts_andor == 'and' ? "(#{search}) and (#{filter})" : "#{search} or (#{filter})" unless filter.blank?
+
+        # taxonomies
```
if Settings[:locations_enabled] do sometihing
if Settings[:organizations_enabled], etc..

Otherwise the migration will stop at this point.

ares replied:
already fixed

dLobatog commented on lib/tasks/migrate_permissions.rake:
```diff
@@ -0,0 +1,207 @@
+desc 'migrate foreman permissions from 1.3 to 1.4'
+task :migrate_permissions => :environment do
+  name = "foreman.#{Time.now.to_i}.sql"
+  `echo changeme | pg_dump -U foreman foreman > #{name}`
```
This will only create a backup in postgres. I'll fix this in my PR for sqlite and mysql

ares replied:
ok so leaving as it is, since it was meant just for testing purposes, thank you

ohadlevy commented on app/assets/javascripts/filters.js:
```diff
@@ -0,0 +1,15 @@
+$(document).ready(function() {
+    $('#filter_resource_type').change(function() {
```
nitpick: use two spaces as indentation.

ares replied:
I set my editor to use 2 spaces for js, it will be fixed in next commit

ohadlevy commented on app/controllers/architectures_controller.rb:
```diff
@@ -3,7 +3,10 @@ class ArchitecturesController < ApplicationController
 
   def index
-    @architectures = Architecture.includes(:operatingsystems).search_for(params[:search], :order => params[:order]).paginate(:page => params[:page])
+    base = Architecture.authorized(:view_architectures)
```
I wonder if we can't meta program this a bit more, as this is really not DRY.

ohadlevy replied:
I would look for examples at https://github.com/josevalim/inherited_resources

ares replied:
refactored, see 1a5aded for more details

ohadlevy commented on app/controllers/concerns/foreman/controller/users_mixin.rb:
```diff
@@ -26,12 +26,6 @@ def clear_params_on_update
     end
   end
 
-  def update_admin_flag
```
why remove from here? doesn't it required for api as well?

ares replied:
you're right, there was another issue that confused me, admin flag setting did not work from CLI because of order of before filters, I'll include fix in next commit, thanks

ohadlevy commented on app/controllers/hosts_controller.rb:
```diff
@@ -35,6 +29,8 @@ def index (title = nil)
         @last_reports = Report.where(:host_id => @hosts.map(&:id)).group(:host_id).maximum(:id)
         # rendering index page for non index page requests (out of sync hosts etc)
+        @authorizer = Authorizer.new(User.current, @hosts)
+        @hostgroup_authorizer = Authorizer.new(User.current, @hosts.map(&:hostgroup).compact.uniq)
```
can we use hostgroup_id instead (just to reduce some extra object creations)

ares replied:
I extended Authorizer and modified controllers where applicable, even though here we need hostgroup included because of views it's a good idea which may be handy later

ohadlevy commented on app/controllers/hosts_controller.rb:
```diff
@@ -386,24 +410,30 @@ def submit_multiple_destroy
   end
 
   def multiple_disable
+    @hosts = find_multiple(:edit_hosts)
```
nitpick: I would prefer all of this calls still to be a method, so if we later on change something small, we change only in one place.

ares replied:
you mean to move @hosts into the find_multiple method? as you probably noticed I'm not fan of that :-)

ares replied:
fixed

ohadlevy commented on app/controllers/lookup_keys_controller.rb:
```diff
@@ -1,22 +1,26 @@
+      values = base.search_for ""
     end
     @lookup_keys = values.includes(:puppetclass).paginate(:page => params[:page])
+    @authorizer  = Authorizer.new(User.current, @lookup_keys)
```
cant authorizer be method on the controller?

ohadlevy commented on app/controllers/puppetclasses_controller.rb:
```diff
@@ -1,21 +1,24 @@
     @host_counter = Host.group(:puppetclass_id).joins(:puppetclasses).where(:puppetclasses => {:id => @puppetclasses.collect(&:id)}).count
     @keys_counter = Puppetclass.joins(:class_params).select('distinct environment_classes.lookup_key_id').group(:name).count
+    @authorizer = Authorizer.new(User.current, @puppetclasses)
+    @hostgroups_authorizer = Authorizer.new(User.current, @puppetclasses.map(&:hostgroups).compact.uniq)
```
ditto

ares replied:
here we can't map just to hostgroup_id because there's n:m relation, Authorizer create will be DRY-ed up later

ohadlevy replied:
you can just query the join table instead :)

ares replied:
fixed (I hope :-)

ohadlevy commented on app/helpers/application_helper.rb:
```diff
@@ -94,25 +94,38 @@ def check_all_roles_links
+  def authorized_for(options)
+    action = options.delete(:auth_action) || options.delete(:action)
+    object = options.delete(:auth_object)
+    user = User.current
```
nitpick: equal sign indentation

ares replied:
fix in following commit

ohadlevy commented on app/models/authorizer.rb:
```diff
@@ -0,0 +1,72 @@
+class Authorizer
```
should this be in app/services?

ares replied:
ok I'll move it

ohadlevy commented on app/models/authorizer.rb:
```diff
@@ -0,0 +1,72 @@
+  attr_accessor :user, :base_collection
+
+  def initialize(user, collection = nil)
+    @cache = Hash.new { |h, k| h[k] = {} }
```
maybe use HashWithInDifference ?

ares replied:
yeah why not

ohadlevy commented on app/models/compute_profile.rb:
```diff
@@ -1,6 +1,5 @@
 class ComputeProfile < ActiveRecord::Base
 
-  include Authorization
```
do we drop auth for this model?

ares replied:
We do for all models, authorization is done on controller level (where we can distinguish between clone and create actions for example). So first checking is done by `ApplicationController#authorize` before_filter that checks whether user has permission for particular action and second takes effect when `.authorized(permission)` scope is called.

However compute profiles were added during work on this PR (actually got there during rebase) so I missed it. I fix permissions for this resource and modified it to use new permission system. (add Authorizable module etc). It will be included in next commit.

ohadlevy commented on app/models/concerns/hostext/search.rb:
```diff
@@ -18,16 +18,21 @@ module Search
       scoped_search :in => :model,       :on => :name,    :complete_value => true, :rename => :model
       scoped_search :in => :hostgroup,   :on => :name,    :complete_value => true, :rename => :hostgroup
       scoped_search :in => :hostgroup,   :on => :label,   :complete_value => true, :rename => :hostgroup_fullname
+      scoped_search :in => :hostgroup,   :on => :id,      :complete_value => true, :rename => :hostgroup_id
```
i dont think we should display all of the id's by default?
e.g.

```
:only_explicit => true
:complete_value => false
```

ares replied:
thanks

ohadlevy commented on app/models/filter.rb:
```diff
@@ -0,0 +1,68 @@
+  scope :limited, lambda { where("search IS NOT NULL") }
+
+  scoped_search :on => :search, :complete_value => true
+  scoped_search :in => :role, :on => :id, :rename => :role
```
why not the other way around? I think users would assume role = name and role_id = num ?

ares replied:
yeah, you're right, the motivation was to resuse ApplicationController#setup_search_options but this is more intuitive so I redefined behaviour to what you say

ohadlevy commented on app/views/filters/_form.html.erb:
```diff
@@ -0,0 +1,68 @@
+      <div class="tab-pane active" id="primary">
+        <% if @role.present? %>
+            <h5><%= _("Selected role: %s" % @role.name) %></h5>
+            <%= f.hidden_field :role_id, :value => f.object.role_id || @role.id %>
```
should it be the other way around?

``` ruby
@role.id || f.object.role_id 
```

ares replied:
I don't think so, @role is present only when you work within its scope. So if you edit a filter that you found on list of filter of a given @role, f.object.role_id should be the same as @role. In some case user could manually modify URL so he enters e.g. `/edit/1?role_id=1` where filter with id 1 has role_id set to 2. User should see current value in form I think therefore 2.

ohadlevy commented on app/views/filters/index.html.erb:
```diff
@@ -0,0 +1,44 @@
+  </tr>
+  </thead>
+  <tbody>
+  <% for filter in @filters %>
```
i cant find the reference online, but afair, using each is better then using for loops

ares replied:
I think it is significant only in partial rendering however I changed it to `.each` syntax. I think I just copied it from another view as a template.

ohadlevy commented on app/views/permissions/index.js.erb:
```diff
@@ -0,0 +1,20 @@
+var permissions = $('#filter_permission_ids');
+permissions.html("<%= escape_javascript(options_for_select(@permissions.map {|p| [p.name, p.id] })) %>");
```
not sure I follow, why dont use a data-attribute instead of dynamic generated js ?

ares replied:
I was not able to modify multiple select properly since I couldn't find any method to change items in API. Also it seems pretty hard to compute search path and determine whether a resource is granular. We'd have to generate a lot of data from rails to keep it consistent. Maybe someone better at javascript could help me with refactor?

ohadlevy commented on app/views/reports/_list.html.erb:
```diff
@@ -14,8 +14,12 @@
+        <% if report.host.nil? %>
+            <td></td>
+        <% else %>
+            <td><%= link_to h(report.host), host_reports_path(report.host) %></td>
```
can drop the h()

ares replied:
hey that's not fair, I haven't touched this line at all and I was fixing an issue I found accidentally :-) (fixed of course, I was kidding)

ohadlevy commented on app/views/roles/report.html.erb:
```diff
@@ -1,3 +1,4 @@
+TODO - OBSOLETE PERMISSION SYSTEM, TO BE REMOVED
```
is that a comment or html?

ares replied:
html, it was removed in previous commit, we don't have permission report anymore (however we could think of adding it in some form after 1.4 if you find it important, personally I have never used this in redmine)

ares commented:
[test]

ares committed:
Seed now creates all roles and permissions

Also some other refactorings and cleanups

jlsherrill commented:
As we discussed on IRC, the Anonymous role was not being created for me in development when i ran db:seed.

Also, i noticed as soon as i went to roles and created a new role, the 'Default user' role appeared out of nowhere.

dLobatog committed:
Remove find_os, leftovers from ares's refactor

DB agnostic backups before permissions migration

thomasmckay commented:
Please add v2 api before merging. This is to facilitate hammer-cli and testing.

jlsherrill commented:
has there been any thought on how you would create a read-only role?  Right now it would be very tedious to do.  Katello created one with all the 'read-only' permissions, could something like that be done here?

ares commented:
@jlsherrill thanks for testing and feedback. Read-only role is created by db:seed and is called 'Viewer'. Basically it's a role with permissions like view_\* grouped by unlimited filters. I agree it's not easy to create it manually and can become hard to manage with new permissions. However we could add some tools to assign multiple permissions easily later (without need of specifying each filter separately).

@thomasmckay as said on IRC, it's out of the scope of this PR, but it's a plan. You can vote for http://projects.theforeman.org/issues/3817

ares committed:
Fixes and tests

ohadlevy commented on lib/tasks/migrate_permissions.rake:
```diff
@@ -0,0 +1,224 @@
+          else
+            raise "could not clear old permissions for role '#{role.name}'"
+          end
+          next
```
next is unused here right?

ohadlevy replied:
oh, it leaves the other loop, ignore me.

ohadlevy commented on lib/tasks/migrate_permissions.rake:
```diff
@@ -0,0 +1,224 @@
+        end
+
+        # check for unknown permissions, this should never happen, raise an exception if it does
+        role_permissions = permission_names.map do |name|
```
this would break on plugin permissions?

ohadlevy replied:
also, what about older plugins that have been removed? I would ignore instead of raise.

dLobatog committed:
GC deferred threshold for tests

This tweak allows us to save some time when running the test suite.
https://37signals.com/svn/posts/2742-the-road-to-faster-tests#all_comments
explains it better.

It went from 11m. 12s, to 8m, 50s, averaging two test runs.

Functional tests adapted to new permission model

Added oj to Gemfile to fix multi_json problem

fix for https://github.com/intridea/multi_json/issues/86 , which raised
when running some tests

dLobatog commented:
Latest changes include test fixes (only 2 integration tests are currently failing) and a patch to speedup our suite

ares committed:
More fixes and tests

dLobatog committed:
Import task for mysql/pg dumps

abenari committed:
move permission migration code into migration

ares committed:
API for plugins adapted to new permissions

ares committed:
Contains:

* DB agnostic backups task
* Split seeds file into separate files

thomasmckay commented:
My notes from a quick test:
    + role name "someone@megacorp.com" invalid, should allow any characters
    + filter json needs to include organizations and locations
    + filter json missing permissions
    + missing api/v2/permissions_controller.rb
    + built in roles (default user and anonymous) should be clickable to see what filters are applied to them

Without full CRUD v2 apis for roles, permissions, and filters, this pull-request is not going to be testable to the degree required. These shouldn't be too much additional code since a lot of the code already exists in the view controllers.

thomasmckay commented:
@ares Could I get a quick set of instructions for having plugin models show up in the authorization? (Specifically I'd like to try getting roles/perms added to a katello engine model.)

ares commented:
@thomasmckay you can see foreman_discovery plugin which adds some new permissions. This PR https://github.com/theforeman/foreman_discovery/pull/49/files adds support for new system. I just had to add resource_type and adapt the controller to use new finders.

So basically you use http://projects.theforeman.org/projects/foreman/wiki/How_to_Create_a_Plugin#Adding-permission but specify a resource type. In that resource you must include Authorizable module and define scoped search (if it's supposed to be granular)

Commit:
v2api - permissions index and show

rbac_v2api - basic v2 api working

rbac_v2api - corrected apipie

thomasmckay commented:
20140206151014_add_taxonomy_searches_to_filter.rb migration errors in foretello (maybe in straight foreman too?)

dLobatog commented:
@thomasmckay Worked fine in Foreman, I tested an upgrade from 1.3 to this on a large db, which error did you get?

ares committed:
Taxonomy support in filters

Plus other fixes and tests

ares commented:
@thomasmckay I encountered it during PR test, seems like it crashes on fresh DB, I will look at it, thank you

ares commented:
@thomasmckay migration fixed

domcleal commented on db/seeds.rb:
```diff
@@ -20,161 +22,34 @@ def audit_modified?(type, name)
-    end
-  end
+# now we load all seed files
+Dir.glob(Rails.root + 'db/seeds.d/*.rb').sort.each do |seed|
```
I like this, but you'll probably need to check the packages work, as the RPMs at least symlink from /var/lib/foreman/db/seeds.rb through to /usr.  Probably needs a spec update.

domcleal replied:
@lzap would you have the time to help here, and just check the upgrade works (with the backups) from a packaged install?

ares replied:
@domcleal what do you think that's needed? I suppose Rails.root should work and since seeds.d is subdir of already listed foreman installation directory I think we don't have to modify the spec. @lzap help with testing package installation would be appreciated

domcleal replied:
The RPM rearranges the 'db' dir slightly strangely, I don't think it'll be packaged at all as it stands without at least a change to include seeds.d, and probably move it up a directory in the spec to /usr/share/foreman/seeds.d/.

domcleal commented on Gemfile:
```diff
@@ -11,11 +11,12 @@ gem 'rest-client', :require => 'rest_client'
 gem 'uuidtools'
 gem "apipie-rails", "~> 0.0.23"
 gem 'rabl', '>= 0.7.5', '<= 0.9.0'
+gem 'oj'
```
For testing?  I don't see any use of it.

ares replied:
good catch, moved to bundler.d/test.rb 
I suppose @eLobato agree that it was for tests only?

dLobatog replied:
Right, I was getting some segfault on my setup that I could only fix with that, probably an issue with my setup.

domcleal commented on app/controllers/api/base_controller.rb:
```diff
@@ -249,5 +260,19 @@ def skip_nested_id
+      when 'index', 'show', 'status'
+        'view'
+      else
+        raise ::Foreman::Exception, "unknown permission for #{params[:controller]}##{params[:action]}"
```
Use string extraction with Foreman::Exception:

```
raise ::Foreman::Exception.new(N_("unknown permission for %s"), "#{params[:controller]}##{params[:action]}")
```

domcleal replied:
I'm a little worried this could fire for actions we've neglected to override action_permission for, and that we might not notice, if our testing or functional tests are insufficient.

Could you perhaps adapt test/lib/foreman/access_permissions_test.rb to verify we don't hit this exception in practice?

ares replied:
I see you concern and I agree but I don't know how to fix it. This method is called only for actions using resource_base / resource_scope inside. Therefore on all others it will fail with this exception. I don't know how to detect whether some action uses this method or not from outside (test), since it's not even before_filter. It does not make sense to me to set some permission name for every action, if they don't use it. If it ever happens it should at least be very easy to find and fix it.

But if you have some idea, I'd be happy to test it somehow.

Note for me, this would work if we were able to filter correct methods:

``` ruby
test "action_permission helper defines permission name for action at #{path}" do
      assert_nothing_raised ::Foreman::Exception do
        controller_instance = controller.new
        controller_instance.stubs(:params).returns({:action => r.defaults[:action]}.with_indifferent_access)
        controller_instance.send :action_permission
      end
end
```

domcleal commented on app/controllers/api/v2/filters_controller.rb:
```diff
@@ -0,0 +1,59 @@
+
+      api :PUT, "/filters/:id/", "Update an filter."
+      param :id, String, :required => true
+      param :filter, Hash, :required => true do
```
Current API style is to use a params group, which saves you duplicating this.  It also means you can make the attrs required when creating and optional when updating.

ares replied:
nice, fixed, thanks!

domcleal commented on app/controllers/api/v2/filters_controller.rb:
```diff
@@ -0,0 +1,59 @@
+        @filters = Filter.authorized(:view_filters).search_for(*search_options).paginate(paginate_options)
+      end
+
+      api :GET, "/filters/:id/", "Show an filter."
```
"a filter", not "an".  ditto below.

domcleal commented on app/controllers/application_controller.rb:
```diff
@@ -124,10 +129,38 @@ def find_by_name
+      when 'index', 'show'
+        'view'
+      else
+        raise ::Foreman::Exception, "unknown permission for #{params[:controller]}##{params[:action]}"
```
ditto re: string extraction

domcleal commented on app/controllers/api/v2/statistics_controller.rb:
```diff
@@ -6,22 +6,22 @@ class StatisticsController < V2::BaseController
-                                            :env_count   => @env_count, :klass_count => @klass_count, :cpu_count => @cpu_count,
-                                            :model_count => @model_count, :mem_size => @mem_size, :mem_free => @mem_free, :swap_size => @swap_size,
-                                            :swap_free   => @swap_free, :mem_totsize => @mem_totsize, :mem_totfree => @mem_totfree }
+        @os_count    = Host.authorized(:view_hosts).my_hosts.count_distribution :operatingsystem
```
It looks like the `my_hosts` scope has been killed off, can this be removed and the scope deleted?

domcleal commented on app/controllers/api/v2/permissions_controller.rb:
```diff
@@ -0,0 +1,32 @@
+        end
+      end
+
+      api :GET, "/permissions/:id/", "Show an permission."
```
'an' again :)

ares replied:
when I created filters API controller I copied base from other controller and gsubbed just resource name, seems like I was not the only one :-) thanks for spotting this

domcleal commented on db/migrate/20140206151014_add_taxonomy_searches_to_filter.rb:
```diff
@@ -0,0 +1,21 @@
+    add_column :filters, :taxonomy_search, :text
+
+    # to precache taxonomy search on all existing filters
+    # setting may not exist yet
```
This should be in app/models/setting/?

ares replied:
I added it to settings model (good idea) but it must also remain here in this form, if you run db:migrate on new database, initializers are executed just once during env load and that one loading default data is skipped if settings table does not exist yet so we don't have default settings at this moment and then calling SETTINGS[:fix_db_cache] = true will fail since the new setting is invalid (missing description etc)

domcleal commented on lib/tasks/backup.rake:
```diff
@@ -0,0 +1,76 @@
+  end
+
+  def postgres_dump(name, config)
+    cmd = "pg_dump #{config['database']} -U #{config['username']} "
```
Nitpick, we recommend `pg_dump -Fc` in the manual.

domcleal commented on lib/tasks/backup.rake:
```diff
@@ -0,0 +1,76 @@
+
+  def sqlite_dump(name, config)
+    cmd = "cp #{config['database']} #{name}"
+    system(cmd)
```
`FileUtils.cp`?

domcleal commented on lib/tasks/backup.rake:
```diff
@@ -0,0 +1,76 @@
+  end
+
+  def mysql_dump(name, config)
+    cmd = "mysqldump #{config['database']} -u #{config['username']} "
```
`mysqldump --opt` highly recommended

domcleal commented on lib/tasks/backup.rake:
```diff
@@ -0,0 +1,76 @@
+  end
+
+  def postgres_import(file)
+    cmd = "psql -d #{config['database']} -U #{config['username']} "
```
pg_restore?

domcleal commented on roles/gh_roles.md:
```diff
@@ -0,0 +1,77 @@
+# Current model
```
docs++

Could you create a docs/ top level dir or similar and file under there?  Just in anticipation of more...

ares replied:
before we start with this - I'm wondering whether there's a better place than code repository for this - redmine wiki? foreman manual? anyway I'll have to update this a bit so I'd like to do it at correct place

domcleal replied:
I'd prefer redmine, I don't think the manual is appropriate for developer level documentation.

domcleal commented on app/views/filters/index.html.erb:
```diff
@@ -0,0 +1,48 @@
+                                                        hash_for_edit_filter_path(:id => filter, :role_id => @role).
+                                                                merge(:auth_object => filter, :authorizer => authorizer))) %>
+        </td>
+        <td><%= filter.resource_type %></td>
```
This isn't very user-friendly.. is it safe to constantize this again and run it back through the translations?  I think we should have friendly translations of model names already.

ares replied:
how do we get translated name of a model? does gettext support `model_name.human`? `_('Architecture')` works and if this is the case, then just wrapping into `_()` should work, because it's basically class name string

ares replied:
fixed

domcleal commented on app/views/filters/_form.html.erb:
```diff
@@ -0,0 +1,68 @@
+
+        <div id="granular_form" style="<%= !f.object.granular? ? 'display:none' : '' %>">
+          <%= checkbox_f f, :unlimited?,
+              :help_inline => _("By unchecking this you can configure filter using scoped_search syntax in search field") %>
```
I dislike mentioning scoped_search specifically, maybe call it Foreman search syntax?  We should migrate the search info to the manual and link it really.

domcleal commented on app/views/filters/_form.html.erb:
```diff
@@ -0,0 +1,68 @@
+            <%= select_f(f, :role_id, Role.scoped, :id, :name) %>
+        <% end %>
+
+        <%= select_f(f, :resource_type, Permission.resources, :to_s, :to_s, { :include_blank => true },
```
ditto re friendliness and i18n of class names here

domcleal replied:
Should include_blank be something like "Everything"?  (If I understand it correctly?)

Edit: I didn't understand correctly, it's actually more of a "Miscellaneous" or "Dashboard" stuff category.

domcleal commented on app/views/filters/_form.html.erb:
```diff
@@ -0,0 +1,68 @@
+
+      <div class="tab-pane active" id="primary">
+        <% if @role.present? %>
+            <h5><%= _("Selected role: %s" % @role.name) %></h5>
```
This looks a bit lost.. maybe use the typical column/field layout to match the UI below it.

domcleal commented on app/views/filters/index.html.erb:
```diff
@@ -0,0 +1,48 @@
+<table class="table table-bordered table-striped">
+  <thead>
+  <tr>
+    <th><%= sort :search, :as => s_("Filter|Search") %></th>
```
It might be clearer to sort columns, so unlimited + search are near each other.
1. Role
2. Resource
3. Unlimited
4. Search
5. Actions

dLobatog committed:
Clone roles

This feature allows users to clone roles easily. Roles are deeply cloned
using the gem 'deep_cloneable', so that the associations (permissions
    and filters) are cloned as well

Fix routes problem on clone role

domcleal commented on app/views/api/v2/filters/main.json.rabl:
```diff
@@ -0,0 +1,17 @@
+
+attributes :search, :role_id, :created_at, :updated_at
+
+child :permissions => :permissions do
```
Child nodes should be in `show` rather than main.

domcleal commented on app/views/api/v2/filters/main.json.rabl:
```diff
@@ -0,0 +1,17 @@
+  attributes :id, :name
+end
+
+child :organizations => :organizations do
```
Wrap with `if SETTINGS[:organizations_enabled]`, ditto below.

domcleal replied:
From #1208, you can use app/views/api/v2/taxonomies/children_nodes.json.rabl here.

ares replied:
will be changed when #1280 get's in, otherwise fixed

domcleal commented on app/views/api/v2/roles/main.json.rabl:
```diff
@@ -2,4 +2,4 @@ object @role
 
 extends "api/v2/roles/base"
 
-attributes :builtin, :permissions, :created_at, :updated_at
```
Filter child node should be added to `show`?

ares replied:
fixed

domcleal commented on config/routes/api/v2.rb:
```diff
@@ -103,6 +103,12 @@
       resources :roles, :except => [:new, :edit]
+      resources :permissions, :only => [:index, :show]
+
+      resources :filters, :except => [:new, :edit] do
```
This should also be nested under roles

thomasmckay commented:
https://github.com/ares/foreman/blob/feature/812_usergroup_roles/app/models/user.rb#L272

Should check for nil roles here? I encountered error on this line while creating a user. Could have just been my test data, though.

thomasmckay commented:
https://github.com/thomasmckay/katello/blob/ak-api-rbac/lib/katello/plugin.rb#L153

This fails on foretello db:migrate

```
$ rake db:migrate
[deprecated] I18n.enforce_available_locales will default to true in the future. If you really want to skip validation of your locale you can set I18n.enforce_available_locales = false to avoid this message.
rake aborted!
PG::UndefinedTable: ERROR:  relation "permissions" does not exist
LINE 5:              WHERE a.attrelid = '"permissions"'::regclass
                                        ^
:             SELECT a.attname, format_type(a.atttypid, a.atttypmod),
                     pg_get_expr(d.adbin, d.adrelid), a.attnotnull, a.atttypid, a.atttypmod
              FROM pg_attribute a LEFT JOIN pg_attrdef d
                ON a.attrelid = d.adrelid AND a.attnum = d.adnum
             WHERE a.attrelid = '"permissions"'::regclass
               AND a.attnum > 0 AND NOT a.attisdropped
             ORDER BY a.attnum
/home/tomckay/.rvm/gems/ruby-1.9.3-p327@foreman/gems/activerecord-3.2.16/lib/active_record/connection_adapters/postgresql_adapter.rb:1163:in `async_exec'
/home/tomckay/.rvm/gems/ruby-1.9.3-p327@foreman/gems/activerecord-3.2.16/lib/active_record/connection_adapters/postgresql_adapter.rb:1163:in `exec_no_cache'
/home/tomckay/.rvm/gems/ruby-1.9.3-p327@foreman/gems/activerecord-3.2.16/lib/active_record/connection_adapters/postgresql_adapter.rb:660:in `block in exec_query'
/home/tomckay/.rvm/gems/ruby-1.9.3-p327@foreman/gems/activerecord-3.2.16/lib/active_record/connection_adapters/abstract_adapter.rb:280:in `block in log'
/home/tomckay/.rvm/gems/ruby-1.9.3-p327@foreman/gems/activesupport-3.2.16/lib/active_support/notifications/instrumenter.rb:20:in `instrument'
/home/tomckay/.rvm/gems/ruby-1.9.3-p327@foreman/gems/activerecord-3.2.16/lib/active_record/connection_adapters/abstract_adapter.rb:275:in `log'
/home/tomckay/.rvm/gems/ruby-1.9.3-p327@foreman/gems/activerecord-3.2.16/lib/active_record/connection_adapters/postgresql_adapter.rb:659:in `exec_query'
/home/tomckay/.rvm/gems/ruby-1.9.3-p327@foreman/gems/activerecord-3.2.16/lib/active_record/connection_adapters/postgresql_adapter.rb:1292:in `column_definitions'
/home/tomckay/.rvm/gems/ruby-1.9.3-p327@foreman/gems/activerecord-3.2.16/lib/active_record/connection_adapters/postgresql_adapter.rb:855:in `columns'
/home/tomckay/.rvm/gems/ruby-1.9.3-p327@foreman/gems/activerecord-3.2.16/lib/active_record/connection_adapters/schema_cache.rb:12:in `block in initialize'
/home/tomckay/.rvm/gems/ruby-1.9.3-p327@foreman/gems/activerecord-3.2.16/lib/active_record/model_schema.rb:229:in `yield'
/home/tomckay/.rvm/gems/ruby-1.9.3-p327@foreman/gems/activerecord-3.2.16/lib/active_record/model_schema.rb:229:in `default'
/home/tomckay/.rvm/gems/ruby-1.9.3-p327@foreman/gems/activerecord-3.2.16/lib/active_record/model_schema.rb:229:in `columns'
/home/tomckay/.rvm/gems/ruby-1.9.3-p327@foreman/gems/activerecord-3.2.16/lib/active_record/model_schema.rb:249:in `column_names'
/home/tomckay/.rvm/gems/ruby-1.9.3-p327@foreman/gems/activerecord-3.2.16/lib/active_record/model_schema.rb:262:in `column_methods_hash'
/home/tomckay/.rvm/gems/ruby-1.9.3-p327@foreman/gems/activerecord-3.2.16/lib/active_record/dynamic_matchers.rb:74:in `all_attributes_exists?'
/home/tomckay/.rvm/gems/ruby-1.9.3-p327@foreman/gems/activerecord-3.2.16/lib/active_record/dynamic_matchers.rb:27:in `method_missing'
/home/tomckay/code/foreman/app/services/foreman/plugin.rb:172:in `permission'
/home/tomckay/code/katello/lib/katello/plugin.rb:154:in `block (2 levels) in <top (required)>'
/home/tomckay/code/foreman/app/services/foreman/plugin.rb:162:in `instance_eval'
/home/tomckay/code/foreman/app/services/foreman/plugin.rb:162:in `security_block'
/home/tomckay/code/katello/lib/katello/plugin.rb:153:in `block in <top (required)>'
/home/tomckay/code/foreman/app/services/foreman/plugin.rb:61:in `instance_eval'
/home/tomckay/code/foreman/app/services/foreman/plugin.rb:61:in `register'
/home/tomckay/code/katello/lib/katello/plugin.rb:1:in `<top (required)>'
/home/tomckay/.rvm/gems/ruby-1.9.3-p327@foreman/gems/polyglot-0.3.3/lib/polyglot.rb:63:in `require'
/home/tomckay/.rvm/gems/ruby-1.9.3-p327@foreman/gems/polyglot-0.3.3/lib/polyglot.rb:63:in `require'
/home/tomckay/.rvm/gems/ruby-1.9.3-p327@foreman/gems/activesupport-3.2.16/lib/active_support/dependencies.rb:251:in `block in require'
/home/tomckay/.rvm/gems/ruby-1.9.3-p327@foreman/gems/activesupport-3.2.16/lib/active_support/dependencies.rb:236:in `load_dependency'
/home/tomckay/.rvm/gems/ruby-1.9.3-p327@foreman/gems/activesupport-3.2.16/lib/active_support/dependencies.rb:251:in `require'
/home/tomckay/code/katello/lib/katello/engine.rb:92:in `block in <class:Engine>'
/home/tomckay/.rvm/gems/ruby-1.9.3-p327@foreman/gems/railties-3.2.16/lib/rails/initializable.rb:30:in `instance_exec'
/home/tomckay/.rvm/gems/ruby-1.9.3-p327@foreman/gems/railties-3.2.16/lib/rails/initializable.rb:30:in `run'
/home/tomckay/.rvm/gems/ruby-1.9.3-p327@foreman/gems/railties-3.2.16/lib/rails/initializable.rb:55:in `block in run_initializers'
/home/tomckay/.rvm/gems/ruby-1.9.3-p327@foreman/gems/railties-3.2.16/lib/rails/initializable.rb:54:in `each'
/home/tomckay/.rvm/gems/ruby-1.9.3-p327@foreman/gems/railties-3.2.16/lib/rails/initializable.rb:54:in `run_initializers'
/home/tomckay/.rvm/gems/ruby-1.9.3-p327@foreman/gems/railties-3.2.16/lib/rails/application.rb:136:in `initialize!'
/home/tomckay/.rvm/gems/ruby-1.9.3-p327@foreman/gems/railties-3.2.16/lib/rails/railtie/configurable.rb:30:in `method_missing'
/home/tomckay/code/foreman/config/environment.rb:5:in `<top (required)>'
/home/tomckay/.rvm/gems/ruby-1.9.3-p327@foreman/gems/polyglot-0.3.3/lib/polyglot.rb:63:in `require'
/home/tomckay/.rvm/gems/ruby-1.9.3-p327@foreman/gems/polyglot-0.3.3/lib/polyglot.rb:63:in `require'
/home/tomckay/.rvm/gems/ruby-1.9.3-p327@foreman/gems/activesupport-3.2.16/lib/active_support/dependencies.rb:251:in `block in require'
/home/tomckay/.rvm/gems/ruby-1.9.3-p327@foreman/gems/activesupport-3.2.16/lib/active_support/dependencies.rb:236:in `load_dependency'
/home/tomckay/.rvm/gems/ruby-1.9.3-p327@foreman/gems/activesupport-3.2.16/lib/active_support/dependencies.rb:251:in `require'
/home/tomckay/.rvm/gems/ruby-1.9.3-p327@foreman/gems/railties-3.2.16/lib/rails/application.rb:103:in `require_environment!'
/home/tomckay/.rvm/gems/ruby-1.9.3-p327@foreman/gems/railties-3.2.16/lib/rails/application.rb:305:in `block (2 levels) in initialize_tasks'
/home/tomckay/.rvm/gems/ruby-1.9.3-p327@foreman/bin/ruby_noexec_wrapper:14:in `eval'
/home/tomckay/.rvm/gems/ruby-1.9.3-p327@foreman/bin/ruby_noexec_wrapper:14:in `<main>'
caused by: (PG::UndefinedTable) ERROR:  relation "permissions" does not exist
LINE 5:              WHERE a.attrelid = '"permissions"'::regclass
                                        ^

    ... skipped 74 linesTasks: TOP => db:migrate => environment
(See full trace by running task with --trace)
```

domcleal commented:
@thomasmckay I hit the same on other plugins, https://github.com/domcleal/foreman/commit/38c0643e9e17f6cb6c3f73578846e10934196fb9 should fix it & other clean migration issues.

ares committed:
Fix for several issues

ares commented:
> https://github.com/ares/foreman/blob/feature/812_usergroup_roles/app/models/user.rb#L272
> 
> Should check for nil roles here? I encountered error on this line while creating a user. Could have just been my test data, though.

@thomasmckay is it still reproducible? I can't see how nil#method error could appear here, backtrace would be helpful

ares commented:
@domcleal I fixed most of your reviews in last commit, I won't rebase until you go through it so you don't have to read it completely again, I'll go over you comments again tomorrow and will mark all that should be fixed, then please let me know when I can rebase

domcleal commented:
@ares thanks, 82720d7 looks good.  Feel free to rebase.

Commit:
fixes #4394 - generate class method on AR collections to retrieve used_taxonomy_ids

lzap commented:
I see this missing dependency:

```
Gem loading error: cannot load such file -- deep_cloneable
```

Solved with:

```
scl enable ruby193 "gem install deep_cloneable"
```

Still testing.

lzap commented:
Migration fails

```
rake aborted!
cannot load such file -- /usr/share/foreman/db/seeds.d/20-permissions
```

Working on that

dLobatog commented:
@lzap I'm currently building `deep_cloneable`, that problem should go away today

lzap commented:
Ok the whole seeds.d/ directory is missing from the distribution. I guess you need to copy those files in this block here:

```
# Put db in %{_localstatedir}/lib/%{name}/db
cp -pr db/migrate db/seeds.rb %{buildroot}%{_datadir}/%{name}
mkdir %{buildroot}%{_localstatedir}/lib/%{name}/db
```

I am heading to lunch, can you fix this and rebase? I will try again after lunch.

lzap commented:
The RPM works fine now.

domcleal commented on app/views/filters/index.html.erb:
```diff
@@ -0,0 +1,48 @@
+                      merge(:auth_object => filter.role, :authorizer => @roles_authorizer))) %>
+            </td>
+        <% end %>
+        <td><%= _(filter.resource_type) %></td>
```
When resource_type is nil, this should output `(Miscellaneous)` to match the form.  At the moment, if you have mark_translated: true in settings.yaml, it blows up as we're passing nil into `_()`, which with debugging enabled, can't handle it.

domcleal commented on app/models/filter.rb:
```diff
@@ -0,0 +1,121 @@
+class Filter < ActiveRecord::Base
```
[fixed] This should include a default_scope with some sane ordering (e.g. by resource_type?).

domcleal commented on app/models/filter.rb:
```diff
@@ -0,0 +1,121 @@
+    self.taxonomy_search = nil if self.taxonomy_search.empty?
+  end
+
+  def parenthize(string)
```
sp: parenthesize :)

GregSutcliffe commented:
If I create a filter like "Parameters / Unlimited / create & edit & destroy" and save, when I try to re-edit the filter I get: http://pastie.org/8780426 - not good.

domcleal commented on app/views/filters/_form.html.erb:
```diff
@@ -0,0 +1,70 @@
+        </div>
+
+        <div id="granular_form" style="<%= !f.object.granular? ? 'display:none' : '' %>">
+          <%= checkbox_f f, :unlimited?,
```
We could improve this in a later PR if you wish, but this UI bothers me somewhat.  I'd like to see the purpose of "Unlimited?" and "Search" to restrict which objects the permissions apply to made clearer on the form, there's currently no explanation of the purpose of these fields.  The word "unlimited" could be quite confusing too in this context, I'd suggest hiding it entirely from the UI and using a friendlier label such as "All resources of this type" or similar.

ares replied:
I think this could be addressed more during work on http://projects.theforeman.org/issues/3807 which should change the UI (not sure about timing though) and hide these in some "advanced" tab

I wish inline help would explain that. If it was too confusing, would improvement of inline help be enough?

domcleal replied:
Yep, add inline help for now then, thanks.

domcleal commented on app/views/filters/_form.html.erb:
```diff
@@ -0,0 +1,70 @@
+      </div>
+
+      <% if show_location_tab? %>
+          <div class="tab-pane" id="locations">
```
Nitpick, indentation

edit: actually, most of this file is incorrectly 4-space

GregSutcliffe commented:
I also can't seem to edit parameters as a user - I've given all rights on Resources LookupKey, Puppet class, Parameter, and HostClass, and yet all edit boxes for parameters (class params or globals) are greyed out. See http://imgur.com/mcseJY3 for the actual result.

domcleal commented on app/models/filter.rb:
```diff
@@ -0,0 +1,121 @@
+  scope :unlimited, lambda { where(:search => nil, :taxonomy_search => nil) }
+  scope :limited, lambda { where("search IS NOT NULL OR taxonomy_search IS NOT NULL") }
+
+  scoped_search :on => :search, :complete_value => true
```
unlimited should also be a searchable field

ehelms commented:
One design conflict I see is the 'filter' model presented here and associated routes has the potential to lead to future conflicts that we may want to address either on this PR or take into account in some Katello work. The assumption here is the dropping of the '/katello' at some point from our routes for a flat route structure. If that is the case, the current content view work that is adding filters will clash as they would create a similar top level route of '/filters' due to the requirement of ensuring all resources have a top level API route. Secondly, there is the potential for model conflicts if we were to ever extend or use the Foreman Filter in our own code. 

@bbuckingham - please look over the above mentioned and give your thoughts as well.

jlsherrill commented:
@ehelms we probably should go through and rename all of our factories to be prepended with katello_filter or something similar.  

The removing of /katello  brings up a good point.  After we do that are all resource paths in katello off limits to use in foreman going forward?  IMHO that is a requirement for the removal of /katello.   Without that guarantee, we probably shouldn't remove /katello.

bbuckingham commented:
@ehelms, @jlsherrill, Correct, if /katello is dropped from the resources, 'filters' will be our first resource naming conflict.  Is there a compelling reason to remove /katello from the resources? 

It seems like it would be confusing to users to name katello 'filters' as 'katello_filters', but to leave all other resources untouched.  We could name that one 'content_view_filters'; however, that might seem a bit redundant if a user chooses to use a nested resource (e.g. GET /content_views/1/content_view_filters/2).

ehelms commented:
I feel like @thomasmckay recently brought up a reason to remove the URL
namespace related to using this PR in Katello. Maybe he can explain.

@jlsherril I agree we might as well do a sweep to change all factories to
be namespaced to address the test failure. I can work on that.

@bbuckingham even with the namespace, I might be inclined to think
/katello/filters are the permission filters but just for Katello resources.
On Feb 25, 2014 4:37 PM, "Brad Buckingham" notifications@github.com wrote:

> @ehelms https://github.com/ehelms, @jlsherrillhttps://github.com/jlsherrill,
> Correct, if /katello is dropped from the resources, 'filters' will be our
> first resource naming conflict. Is there a compelling reason to remove
> /katello from the resources?
> 
> It seems like it would be confusing to users to name katello 'filters' as
> 'katello_filters', but to leave all other resources untouched. We could
> name that one 'content_view_filters'; however, that might seem a bit
> redundant if a user chooses to use a nested resource (e.g. GET
> /content_views/1/content_view_filters/2).
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/theforeman/foreman/pull/1070#issuecomment-36061235
> .

jlsherrill commented:
@bbuckingham for content view versions we use  content_view_version for  the model name and route, so it might be as consistent to do content_view_filter

I wasn't meaning to name the filters 'katello_filters', only naming all of our factories with the katello_ prefix.

domcleal commented:
Namespacing factories sounds good, thanks.

Regarding routes, I'd prefer if filters (in Foreman) was nested under /roles (similar to /compute_resources/:id/images), and in Katello it sounds like it should be nested under content_view.

thomasmckay commented:
@ehelms re: /katello routes prefix, I think there is a solution that. Here is the code that caused problems:
https://github.com/thomasmckay/foreman/blob/294acea831c4145a7cfb6ca0a3228631887190f2/app/controllers/permissions_controller.rb#L28

and what @ares figured would work

```
      object_path = f.object.resource_type.pluralize.underscore
      if object_path.include?('/')
        prefix, suffix = object_path.split('/', 2)
        send(prefix).send(suffix + '_path')
      else
        send("#{object_path}_path")
      end
```

I'd wait to implement this until a later pull-request when we begin adding new rbac to katello side.

ares commented:
> Regarding routes, I'd prefer if filters (in Foreman) was nested under /roles (similar to 
> /compute_resources/:id/images), and in Katello it sounds like it should be nested under 
> content_view.

You can also work with filters of more than one role at the same time. Therefore in that case we require `/filters` and `/roles/:id/filters` so it wouldn't help much.

bbuckingham commented:
@jlsherrill, @ehelms, Once I finish the publish changes, I'll work on updates to the cv-rework branch to rename filters with the prepended 'content_views'.  Will do this for both filters and the related filter rules.

ares commented:
@GregSutcliffe both issues have been fixed, could you please retest?

ehelms commented:
@jlsherril @bbuckingham the model level makes sense to me but doing the
routes that way feels overly verbose. But I wonder if top level filters
make sense for content views if I cannot associate a filter to other CVS. I
know it's partially a convention but still gives me pause.
On Feb 26, 2014 8:03 AM, "Brad Buckingham" notifications@github.com wrote:

> @jlsherrill https://github.com/jlsherrill, @ehelmshttps://github.com/ehelms,
> Once I finish the publish changes, I'll work on updates to the cv-rework
> branch to rename filters with the prepended 'content_views'. Will do this
> for both filters and the related filter rules.
> 
> ## 
> 
> Reply to this email directly or view it on GitHubhttps://github.com/theforeman/foreman/pull/1070#issuecomment-36122296
> .

jlsherrill commented:
@ehelms @bbuckingham so for versions we use:

  /content_views/:id/versions
for listing for a particular view, but:

/content_view_versions/    for all other things.  Something similar could be done with filters?

bbuckingham commented:
@jlsherrill, I like that approach for the routes.  Unless there are objections, will do that.

GregSutcliffe committed:
Make host#edit use consistent field rules

GregSutcliffe commented:
@ares confirmed both parameter issues fixed. I've sent you a PR for the scoping issues on domains and hostgroups. I can't find any other problems, so I think I'm good with this :+1:

domcleal commented on app/models/filter.rb:
```diff
@@ -0,0 +1,134 @@
+  has_many :filterings, :dependent => :destroy
+  has_many :permissions, :through => :filterings
+
+  default_scope lambda { order('role_id') }
```
role_id isn't unique, maybe add 'id' as the second order parameter?

domcleal commented on app/controllers/roles_controller.rb:
```diff
@@ -37,12 +37,21 @@ def create
+    new_role = @role.dup :include => [:filters => :permissions]
+    new_role.name += '_clone'
+    if new_role.save
+      flash[:notice] = "Role #{new_role.name} cloned from role #{@role.name}"
```
Missing i18n extraction, ditto below.

domcleal commented on app/controllers/roles_controller.rb:
```diff
@@ -37,12 +37,21 @@ def create
 
+  def clone
+    new_role = @role.dup :include => [:filters => :permissions]
+    new_role.name += '_clone'
```
We have a different approach to cloning of both hosts and host groups, instead we .dup in memory and render the "new" form.  The name is set to "" so the user is prompted to fill one in.  This way you also don't need error handling for .save below.

ares replied:
@eLobato would you mind modifying cloning this way?

domcleal commented on app/controllers/hosts_controller.rb:
```diff
@@ -528,8 +533,41 @@ def template_used
+
+  def action_permission
+    case params[:action]
+      when 'clone', 'externalNodes', 'bmc', 'pxe_config', 'storeconfig_klasses',
```
For clone, I think this should be :create

ares replied:
I think it refers to view the original, but you're right, we should probably check somehow for two permissions here, maybe just check whether `User.current.can?(:create_roles)`
@eLobato?

edit: sorry I thought it was related to role cloning, I'll go over all cloning actions and will add create permission checks myself

ares replied:
So only action where we should check for create permission is for roles cloning, which will probably be modified. All view permissions are correct since they are used for searching the original that we are going to clone. Since clone just prepare new object, we check for create permission in create action.

@domcleal does this make sense?

domcleal replied:
I see your reasoning, but I'm guessing you wouldn't show the link to create a "new role" if the user didn't have the create permission, so we also shouldn't show a link to clone roles if the user isn't going to be able to create it.  Cloning isn't a substitute for viewing.

ares replied:
Ah I understand the source of confusion now. For displaying and allowing to perform some action we use permission map defined in `app/services/foreman/access_permissions.rb` where clone actions are allowed by create_\* permissions.

`action_permission` method you're referring to is used to authorize resource objects collection within this action. It say what permission should be used to search objects if user has been already allowed to perform that action (based on permissions map, not this permission).

So if users do not have create permission they do not see clone button/link and they would not be allowed to even trigger this action. View permission could probably fail only if they have create permission and view permission for only some objects. They could modify id of link they see (or forges url any other way).

domcleal replied:
Gotcha, thanks for the explanation!

domcleal commented on app/controllers/hostgroups_controller.rb:
```diff
@@ -140,4 +133,13 @@ def subscribed_users
+  def action_permission
+    case params[:action]
+      when 'nest', 'clone'
+        'view'
```
create?

domcleal commented on app/controllers/hosts_controller.rb:
```diff
@@ -464,7 +469,7 @@ def disabled
 
   def process_hostgroup
-    @hostgroup = Hostgroup.find(params[:host][:hostgroup_id]) if params[:host][:hostgroup_id].to_i > 0
+    @hostgroup = Hostgroup.authorized(:view_hostgroups).find(params[:host][:hostgroup_id]) if params[:host][:hostgroup_id].to_i > 0
```
In light of 56285d8, shouldn't these process_\* calls skip calling authorized?

domcleal commented on app/controllers/roles_controller.rb:
```diff
@@ -59,16 +67,18 @@ def destroy
+  def action_permission
+    case params[:action]
+      when 'clone'
+        'view'
```
create?

domcleal commented on app/views/filters/index.html.erb:
```diff
@@ -0,0 +1,48 @@
+                  merge(:auth_object => filter.role, :authorizer => @roles_authorizer))) %>
+        </td>
+      <% end %>
+      <td><%= _(filter.resource_type || '(Miscellaneous)') %></td>
```
Correction, I think this should be:

```
<%= _(filter.resource_class.try(:humanize_class_name) || filter.resource_type || '(Miscellaneous)') %>
```

Though it exposes some bugs in a) the model extractions, which seem to not be adding namespaced stuff (e.g. Host::Managed, audits), and b) we're missing some English translations.  Both of those can be fixed though.

EDIT: filed as 4502, 4503.

domcleal commented:
@ehelms are you close to fixing the factory issue?  We're looking to merge this today or very early next week.

ares committed:
Reviews

domcleal commented on app/models/permission.rb:
```diff
@@ -0,0 +1,20 @@
+  end
+
+  def self.with_translations
+    resources.map { |r| [_(r), r] }
```
This will need the same fix as the role index, converting the resource_type to class and getting the humanized name, so it shows up properly on the filter edit page.

ehelms commented:
@domcleal running through the test suite as we speak

Commit:
Adds deep cloneable dependency

domcleal commented:
[test]

dLobatog committed:
Clone role through Roles#new form

Added Roles#create test

domcleal commented on app/controllers/roles_controller.rb:
```diff
@@ -59,16 +66,31 @@ def destroy
+  def role_from_form
+    if params[:original_role_id].present?
+      new_role = Role.find(params[:original_role_id]).
+                   dup(:include => [:filters => :filterings])
```
The "builtin" attribute is also getting copied here, so if you clone "Default user" then you end up with a role that you can't delete or rename.

domcleal commented on app/controllers/roles_controller.rb:
```diff
@@ -59,16 +66,31 @@ def destroy
+  def action_permission
+    case params[:action]
+      when 'clone'
+        'create'
```
'view'

ares committed:
Fix cloning reviews

ohadlevy commented on app/controllers/api/v2/smart_proxies_controller.rb:
```diff
@@ -63,8 +63,17 @@ def refresh
+        return SmartProxy.authorized(:view_smart_proxies).includes(:features).try(type.downcase+"_proxies") if type.present?
+        return SmartProxy.authorized(:view_smart_proxies).includes(:features).scoped
+      end
+      
```
nitpick: extra 6 whitespace

ohadlevy commented on app/models/operatingsystems/windows.rb:
```diff
@@ -12,6 +12,10 @@ def pxedir
   def url_for_boot(file)
     pxedir + "/" + PXEFILES[file]
   end
+  
```
nitpick: extra whitespace too :)

ares committed:
Fix autocompletion errors in edit form

domcleal commented:
Merged as acfbc45 for Foreman 1.5.0.  Thanks @ares et al for the hard work!
